pkg/cvo/updatepayload.go

```diff
@@ -162,9 +164,9 @@ func (r *payloadRetriever) targetUpdatePayloadDir(ctx context.Context, update co
 	payloadHash := base64.RawURLEncoding.EncodeToString(hash.Sum(nil))
 	tdir := filepath.Join(r.workingDir, payloadHash)
 
-	// Prune older jobs and directories while gracefully handling errors.
-	if err := r.pruneJobs(ctx, 0); err != nil {
-		klog.Warningf("failed to prune jobs: %v", err)
+	// Prune older pods and directories while gracefully handling errors.
+	if err := r.prunePods(ctx); err != nil {
+		klog.Warningf("failed to prune pods: %v", err)
```

Comment:
I honestly do not know the answer.

Let me understand the concern:
- CVO started Pod A retrieving the payload.
- CVO got restarted (while Pod A is still running) and started to prune Pod A but failed for some unknown reason.
- CVO started Pod B retrieving the payload.
Then Pod A and Pod B may "clobber each others output"?

Would this be the same story of racing if "Pod"s are replaced with "Job"s above?
The Job controller may fail on deleting the pod for the job too.

Reply:
Would this be the same story of racing if "Pod"s are replaced with "Job"s above?

https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs

I think i get the answer, "normally, only one Pod is started, unless the Pod fails." implies that with Job, Pod for Job B wont get started when Pod for Job A is still running.

Reply:
The concern is mainly whether it is safe to continue at L169 - we know that `r.prunePods` failed, for whatever reason (`err != nil`). We log it (correct), but then we continue. It can be something like "apiserver was disrupted at that very moment"  or something like that. 

Is it safe to continue working? As opposed to returning and maybe trying the whole procedure again later? By "clobbering the output" I meant  whether two simultaneously executing pods (one that we failed to delete, and one that we will spawn after we continue) are somehow not conflicting. The `version` pods' job is to extract the manifests into the filesystem shared with CVO itself - hypothetically, the two pods could write the pods into the same location (maybe, I did not look at the actual code).

Reply:
It looks like continuing on error here was me in 507b474632 (pkg/cvo/updatepayload: Prune previous payload downloads, 2022-04-03, #760) when we went from "no pruning at all" to "best-effort pruning".  I'm comfortable moving to "require pruning" and failing out of this function on error instead of just warning about errors.  But I agree with Hongkai that it seems like preexisting exposure.